### PR TITLE
feature: support manipulate headers with variables

### PR DESCRIPTION
**Pull request by MengJiapeng**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
通过路由规则新增/修改头部时，支持使用变量动态设置值；变量命名规则同accesslog的format表达式

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by MengJiapeng** [comment]
以下为自动回复    您好，您的邮件我已收到，我会尽快给您回复！

File: pkg/router/headerformatter.go [hunk 1]
```diff
@@ -47,6 +66,73 @@ func (f *plainHeaderFormatter) append() bool {
...
+	} else if strings.HasPrefix(v.variableName, types.VarPrefixRespHeader) {
+		prefix = types.VarPrefixRespHeader
+		isPrefix = true
+	}
+	if isPrefix {
+		headerName := strings.TrimPrefix(v.variableName, prefix)
+		ret, _ := headers.Get(headerName)
+		return ret
+	}
+	switch v.variableName {
```
Comment (taoyuanyuan) [inline pkg/router/headerformatter.go:95]: we should call `variable.GetString(ctx, variableName)`, to support any variable.

File: pkg/router/types.go [hunk 2]
```diff
@@ -50,7 +50,7 @@ var (
 )
 
 type headerFormatter interface {
-	format(requestInfo api.RequestInfo) string
+	format(headers types.HeaderMap, requestInfo api.RequestInfo) string
```
Comment (taoyuanyuan) [inline pkg/router/types.go:53]: I think `headers` and `requestInfo` are not needed,  just `ctx` to support variable..

File: pkg/router/headerformatter.go [hunk 3]
```diff
@@ -47,6 +66,73 @@ func (f *plainHeaderFormatter) append() bool {
...
+	)
+	if strings.HasPrefix(v.variableName, types.VarPrefixReqHeader) {
+		prefix = types.VarPrefixReqHeader
+		isPrefix = true
+	} else if strings.HasPrefix(v.variableName, types.VarPrefixRespHeader) {
+		prefix = types.VarPrefixRespHeader
+		isPrefix = true
+	}
+	if isPrefix {
+		headerName := strings.TrimPrefix(v.variableName, prefix)
```
Comment (doujiang24) [inline pkg/router/headerformatter.go:91]: maybe we should move these lines to `getHeaderFormatter`? for better performance.

File: pkg/router/headerformatter.go (hunk 1 above)
Comment (MengJiapeng) [reply pkg/router/headerformatter.go:95]: got it! maybe I should understand how variable works first.

File: pkg/router/types.go (hunk 2 above)
Comment (MengJiapeng) [reply pkg/router/types.go:53]: It seems that we didn't set the `requestInfo` into the `ctx`. And we set the request headers and response headers with different context key, but in the `format` method, we can't determine which header we are handing now,  so I think the parameters are necessary here.

File: pkg/router/headerformatter.go (hunk 1 above)
Comment (taoyuanyuan) [reply pkg/router/headerformatter.go:95]: The variables are global in MOSN, we can get them in different modules of a request lifecycle.

File: pkg/router/headerformatter.go (hunk 3 above)
Comment (MengJiapeng) [reply pkg/router/headerformatter.go:91]: yeah, I will add a new headerformatter to deal with this situation.

File: pkg/router/types.go (hunk 2 above)
Comment (MengJiapeng) [reply pkg/router/types.go:53]: It seems that with `variable` api, we can get any value with `ctx`, I will try later!

File: pkg/router/headerformatter.go [hunk 4]
```diff
@@ -18,17 +18,36 @@
...
 
 	"mosn.io/mosn/pkg/log"
-	"mosn.io/mosn/pkg/types"
 )
 
+var variableValueRegexp = regexp.MustCompile(`^%.+%$`)
+
 func getHeaderFormatter(value string, append bool) headerFormatter {
-	// TODO: variable headers would be support very soon
+	if variableValueRegexp.MatchString(value) {
```
Comment (taoyuanyuan) [inline pkg/router/headerformatter.go:32]: maybe we don't need introduce regex, and it's easy to judge that first and last string are `%`.

Comment (MengJiapeng) [reply pkg/router/headerformatter.go:32]: Benchmark shows that `strings.HasPrefix` and `strings.HasSuffix` are faster than regex......I will modify it later.
Another question is that if we got a value which not begins and ends with `%`, but contains '%', should we ignore it and return nil or treat it as a plain value?

Comment (taoyuanyuan) [reply pkg/router/headerformatter.go:32]: I think we should treat it as a plain value, beacause we do not prohibit `%` as a part of header

File: pkg/router/headerformatter.go [hunk 5]
```diff
@@ -18,19 +18,21 @@
...
 )
 
 func getHeaderFormatter(value string, append bool) headerFormatter {
-	// TODO: variable headers would be support very soon
-	if strings.Index(value, "%") != -1 {
-		if log.DefaultLogger.GetLogLevel() >= log.WARN {
-			log.DefaultLogger.Warnf("variable headers not support yet, skip, value: %s", value)
+	if len(value) > 2 && strings.HasPrefix(value, "%") && strings.HasSuffix(value, "%") {
+		variableName := strings.Trim(value, "%")
+		// todo cache the variable so we don't need to find it in format method
```
Comment (taoyuanyuan) [inline pkg/router/headerformatter.go:29]: delete expired  todo comment.

Comment (MengJiapeng) [reply pkg/router/headerformatter.go:29]: todo comment here is not expired. A better implementation can be that we save the `variable.Check` returned `variable` into the variableHeaderFormatter, so we can just get the value in `format` method. But currently, the `variable` api dosen't provide a proper method to get value from variable, so we use `variable.GetString` method which find value from all registered variables , which can be optimized.

Comment (taoyuanyuan) [reply pkg/router/headerformatter.go:29]: oh～ Good idea！

**Comment by taoyuanyuan** [comment]
thanks for your contribution～

**Review by taoyuanyuan — APPROVED** [review]

**Review by doujiang24 — APPROVED** [review]

**Pull request merged**